src/PostgREST/App.hs

```diff
@@ -205,8 +204,36 @@ app dbStructure proc conf apiRequest =
                     then responseLBS s [toHeader contentType, r] (toS body)
                     else responseLBS s [r] ""
 
-        (ActionDelete, TargetIdent _, Nothing) ->
-          case mutateSqlParts of
+        (ActionSingleUpsert, TargetIdent (QualifiedIdentifier tSchema tName), Just (PayloadJSON payload pType pKeys)) ->
+          case mutateSqlParts tSchema tName of
+            Left errorResponse -> return errorResponse
+            Right (sq, mq) -> do
+              let isSingle = case pType of
+                               PJArray len -> len == 1
+                               PJObject -> True
+                  colNames = colName <$> tableCols dbStructure tSchema tName
+              if topLevelRange /= allRange
+                then return $ simpleError status400 [] "Range header and limit/offset querystring parameters are not allowed for PUT"
+              else if not isSingle
+                then return $ simpleError status405 [] "PUT payload must contain a single row"
+              else if S.fromList colNames /= pKeys
+                then return $ simpleError status405 [] "You must specify all columns in the payload when using PUT"
+              else do
+                row <- H.query (toS payload) $
+                       createWriteStatement sq mq (contentType == CTSingularJSON) False
+                                            (contentType == CTTextCSV) (iPreferRepresentation apiRequest) []
+                let (_, queryTotal, _, body) = extractQueryResult row
+                if queryTotal /= 1
+                  then do
+                    HT.condemn
+                    return $ simpleError status405 [] "The URI must univocally identify the payload when using PUT"
```

Comment:
This error message might be unclear. What about, "PUT must be applied to exactly one resource." Maybe include a hint like, "Request attempted to modify ${queryTotal} table rows"

Reply:
Is not clear from the code but I'm enforcing a different condition here, so first of all my interpretation of [RFC 7231](https://tools.ietf.org/html/rfc7231#section-4.3.4) is that the PUT uri must identify the payload being inserted so that the next GET can return the same exact payload right? So what I'm doing here is this:

```haskell
-- table numbers is empty
it "fails if the uri doesn't identify the payload" $
  put "/numbers?id=eq.7" [str| [ { "id": 11, "name": "eleven"} ]|] `shouldRespondWith` 405
```

This is actually done by the [sql query](https://github.com/steve-chavez/postgrest/blob/upsert/src/PostgREST/QueryBuilder.hs#L290) that WHEREs the payload according to the querystring parameters, if that WHERE gives false then nothing gets upserted and the `queryTotal` will be 0.

Reply:
Check my comments in https://github.com/begriffs/postgrest/pull/1048#discussion_r163925372, if I'm misunderstanding something in the RFC I'll change it but for now not sure what would be the clearer message.

Reply:
I'm haven't figured out how to make a request that causes this error message to happen. What do I need to do?

Reply:
This test shows how to make the error happen: https://github.com/begriffs/postgrest/pull/1048/files#diff-385e57e2f6697b6568f08e4d151b1d7cR133

Reply:
Thanks, now I see. What about adding a hint to the error message, like

```json
{
  "message": "The URI must univocally identify the payload when using PUT",
  "hint": "Changing primary key values with PUT is not allowed"
}
```

The existing message is fine, but the hint could make it more concrete for people.

Reply:
Or `"hint": "Payload values do not match resource in primary key column(s)"`
whatever you think explains it best.